### PR TITLE
fix py save packet memleak

### PR DESCRIPTION
**Pull request by pstorz**
#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General

- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [ ] Separate commit for this PR in the CHANGELOG.md, PR number referenced is same
- [x] Commit descriptions are understandable and well formatted

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `check-sources --since-merge` does not report any problems
- [x] `git status` should not report modifications in the source tree after building and testing

### CODE REVIEW

File: core/src/lib/message_queue_item.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ class MessageQueueItem {
   dlink link_;
   int type_ = 0;
   utime_t mtime_ = {0};
-  std::string msg_;
+  std::string* msg_{nullptr};
```
Comment (franku) [inline core/src/lib/message_queue_item.h:38]:
```suggestion
  const char* msg_{nullptr};
```

File: core/src/lib/message.cc [hunk 2]
```diff
@@ -1615,18 +1615,18 @@ void Qmsg(JobControlRecord* jcr, int type, utime_t mtime, const char* fmt, ...)
   }
 
   item = (MessageQueueItem*)malloc(sizeof(MessageQueueItem));
-  new (item) MessageQueueItem();
   item->type_ = type;
   item->mtime_ = time(NULL);
-  item->msg_ = buf.c_str();
+  item->msg_ = new std::string(buf.c_str());
```
Comment (franku) [inline core/src/lib/message.cc:1620]:
```suggestion
  item->msg_ = strdup(buf.c_str());
```

File: core/src/lib/message.cc [hunk 3]
```diff
@@ -1615,18 +1615,18 @@ void Qmsg(JobControlRecord* jcr, int type, utime_t mtime, const char* fmt, ...)
...
 
   if (!jcr) { jcr = GetJcrFromThreadSpecificData(); }
 
   /*
    * If no jcr  or no JobId or no queue or dequeuing send to syslog
    */
   if (!jcr || !jcr->JobId || !jcr->msg_queue || jcr->dequeuing_msgs) {
-    syslog(LOG_DAEMON | LOG_ERR, "%s", item->msg_.c_str());
+    syslog(LOG_DAEMON | LOG_ERR, "%s", item->msg_->c_str());
```
Comment (franku) [inline core/src/lib/message.cc:1628]:
```suggestion
    syslog(LOG_DAEMON | LOG_ERR, "%s", item->msg_);
```

File: core/src/lib/message.cc [hunk 4]
```diff
@@ -1615,18 +1615,18 @@ void Qmsg(JobControlRecord* jcr, int type, utime_t mtime, const char* fmt, ...)
...
+    delete item->msg_;
```
Comment (franku) [inline core/src/lib/message.cc:1629]:
```suggestion
    free(item->msg_);
    item->msg_ = nullptr;
```

File: core/src/lib/message.cc [hunk 5]
```diff
@@ -1650,7 +1650,8 @@ void DequeueMessages(JobControlRecord* jcr)
   P(jcr->msg_queue_mutex);
   jcr->dequeuing_msgs = true;
   foreach_dlist (item, jcr->msg_queue) {
-    Jmsg(jcr, item->type_, item->mtime_, "%s", item->msg_.c_str());
+    Jmsg(jcr, item->type_, item->mtime_, "%s", item->msg_->c_str());
+    delete (item->msg_);
```
Comment (franku) [inline core/src/lib/message.cc:1654]:
```suggestion
    free(item->msg_);
    item->msg_ = nullptr;
```

**Review by franku — COMMENTED** [review]
Would suggest to use const char* and strdup because it is not plausible to use a std::string on the heap here.

**Review by sduehr — CHANGES_REQUESTED** [review]
Looks good now, ready for merge, only CHANGELOG.md entry is missing.

File: core/src/lib/message.cc [hunk 6]
```diff
@@ -1559,16 +1559,17 @@ void Qmsg(JobControlRecord* jcr, int type, utime_t mtime, const char* fmt, ...)
...
-  item->msg_ = buf.c_str();
+  item->msg_ = strdup(buf.c_str());
 
   if (!jcr) { jcr = GetJcrFromThreadSpecificData(); }
 
   // If no jcr  or no JobId or no queue or dequeuing send to syslog
   if (!jcr || !jcr->JobId || !jcr->msg_queue || jcr->dequeuing_msgs) {
-    syslog(LOG_DAEMON | LOG_ERR, "%s", item->msg_.c_str());
+    syslog(LOG_DAEMON | LOG_ERR, "%s", item->msg_);
+    free(const_cast<char*>(item->msg_));
```
Comment (arogge) [inline core/src/lib/message.cc:1571]:
```suggestion
    free(item->msg_);
```

File: core/src/lib/message.cc [hunk 7]
```diff
@@ -1588,7 +1589,9 @@ void DequeueMessages(JobControlRecord* jcr)
   P(jcr->msg_queue_mutex);
   jcr->dequeuing_msgs = true;
   foreach_dlist (item, jcr->msg_queue) {
-    Jmsg(jcr, item->type_, item->mtime_, "%s", item->msg_.c_str());
+    Jmsg(jcr, item->type_, item->mtime_, "%s", item->msg_);
+    free(const_cast<char*>(item->msg_));
```
Comment (arogge) [inline core/src/lib/message.cc:1593]:
```suggestion
    free(item->msg_);
```

File: core/src/lib/message_queue_item.h [hunk 8]
```diff
@@ -35,7 +35,7 @@ class MessageQueueItem {
   dlink link_;
   int type_ = 0;
   utime_t mtime_ = {0};
-  std::string msg_;
+  const char* msg_{nullptr};
```
Comment (arogge) [inline core/src/lib/message_queue_item.h:38]:
```suggestion
  char* msg_{nullptr};
```

**Review by arogge — CHANGES_REQUESTED** [review]
If it isn't `const` to us, we shouldn't delcare it `const`
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const

**Pull request merged**